### PR TITLE
Added LOO CV optimization of kernel hyperparameters

### PR DESCRIPTION
**Pull request by costashatz**
This PR adds Leave One Out Cross Validation for the optimization of the hyperparameters (see #221). I added some regression test for the gradient computations. The gradient computation can be improved regarding the performance. At the moment, LOO-CV is slower than the Maximum Likelihood Estimation (MLE)... Overall, LOO-CV and MLE work more or less the same, with MLE being a bit more robust in smooth functions wheres LOO-CV is more robust in discontinuous ones (e.g., step-like or linear piece-wise ones)..

Coding it, I realized that we should change a bit how we have the optimization of the hyper-parameters. I think we should make the following functions, member functions of the GP class and the optimization procedures should only call them:

- `{double, Eigen::VectorXd} compute_log_lik(bool compute_gradient = false) const`
- `{double Eigen::VectorXd} compute_log_loo_cv(bool compute_gradient = false) const`

This will allows us to have the computations only in one place (less error-prone) and have less re-writing..

What do you think? If you agree, I can make the changes in this PR...

### CODE REVIEW

**Comment by costashatz** [comment]
@jbmouret @Aneoshun I made an effort to put all the computations of log likelihood and log probability of LOO-CV into the GP. The code in the HP optimizations looks much simpler and we do not have to re-write the same code... Let me know if you like it..

**Comment by costashatz** [comment]
@Aneoshun any feedback on this? Thanks!

**Comment by Aneoshun** [comment]
Hi @costashatz,

I like your efforts to put the Lik, and CV computation inside the GP. 

One question: When you compute the LOO-CV in a "colwise" fashion. If I am correct, you only need the colwise when you compute this: _alpha.array().square().array().colwise()
You can probably improve this by computing the sum at the begining and then only relying on vectors, which should be slightly more efficient):

_log_loo_cv = (((-0.5 * (_alpha.array().square().array().colwise().sum() * inv_diag.array())).array() - 0.5 * inv_diag.array().log().array()) - 0.5 * std::log(2 * M_PI)).sum();

By the way, are you sure about this part:
((-0.5 * (_alpha.array().square().array().colwise() * inv_diag.array())).array().colwise() - 0.5 * inv_diag.array().log().array()
I am wondering, does the ".colwise() - ....." add  "...." on all the elements of each columns? It looks to me that this expression actually multiply by the number of samples the log part.  

I did not look the Grad computation part, but I trust your unittests.

**Comment by costashatz** [comment]
> I like your efforts to put the Lik, and CV computation inside the GP.

Thanks...

> One question: When you compute the LOO-CV in a "colwise" fashion. If I am correct, you only need the colwise when you compute this: _alpha.array().square().array().colwise()
> You can probably improve this by computing the sum at the begining and then only relying on vectors, which should be slightly more efficient):
> 
> _log_loo_cv = (((-0.5 * (_alpha.array().square().array().colwise().sum() * inv_diag.array())).array() - 0.5 * inv_diag.array().log().array()) - 0.5 * std::log(2 * M_PI)).sum();

If we assume we have 1-D outputs only, then we can compute in a bit more efficient way (we remove all the colwise and we have just to sum rowwise). On the other hand, to properly compute the values with multiple outputs, each column is a different output dimension. As such, `inv_diag` and `_alpha.array().square().array().colwise().sum()` is not of the same dimension: the first one is Nx1, whereas the second one is 1xDIM_OUT. Basically, I need to add and divide the 2 constant (the one not so constant, but constant colwise) values colwise, then do a sum rowwise (you add the log-probabilities of removing each sample -- check the equations in the GP book, page 116) and in the end I add the values of each output dimension (colwise sum). That's why is a mess. In short we want the following:

```cpp
loo_cv = loo_cv_out_0 + loo_cv_out_1 + ... + loo_cv_out_d;
```
where:

```
loo_cv_out_i = Sum(j from 0 to N-1)[log-prob of removing j-th sample], for the i-th output dimension
```

I didn't find a better way to do it, but if you have any ideas feel free (the one you gave now, it's not working --- I tried to make it as it should and got back my original one).

> By the way, are you sure about this part:
> ((-0.5 * (_alpha.array().square().array().colwise() * inv_diag.array())).array().colwise() - 0.5 * inv_diag.array().log().array()
> I am wondering, does the ".colwise() - ....." add "...." on all the elements of each columns? It looks to me that this expression actually multiply by the number of samples the log part.

If you check [here](https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html) (the part on broadcasting), I think I am using it correctly. If you feel there's an issue, you can make some easy checks (I didn't find anything).

**Comment by Aneoshun** [comment]
oups, I did not take into consideration the multi-dimensional aspects... My bad. 
Everything makes sense now. Sorry for my questions.

**Comment by costashatz** [comment]
@Aneoshun Thanks for the feedback.. 

@jbmouret any thoughts?

**Comment by jbmouret** [comment]
**OK for me** (at first sight)

**Comment by jbmouret** [comment]
Should we merge? to 'experimental' maybe?

**Comment by costashatz** [comment]
> Should we merge? to 'experimental' maybe?

I think we should merge into master. It doesn't break the previous code and I believe is more compact code. LOO-CV may have some small bug, but finite differences check is passing so it's highly unlikely that it will change its behavior. We could improve it in time though (but we can open a PR when we need it).

**Comment by costashatz** [comment]
@jbmouret so do I merge?
